cli/permissions.rs

```diff
@@ -354,3 +421,137 @@ fn permission_prompt(message: &str) -> DenoResult<PromptResult> {
     };
   }
 }
+
+fn check_path_white_list(
+  filename: &str,
+  white_list: &Arc<HashSet<String>>,
+) -> bool {
+  let mut path_buf = PathBuf::from(filename);
+
+  if white_list.contains(path_buf.to_str().unwrap()) {
+    return true;
+  }
+
+  while path_buf.pop() {
+    if white_list.contains(path_buf.to_str().unwrap()) {
+      return true;
+    }
+  }
+  false
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+
+  #[test]
+  fn check_paths() {
+    let whitelist = vec![
+      "/a/specific/dir/name".to_string(),
+      "/a/specific".to_string(),
+      "/b/c".to_string(),
+    ].to_vec();
+
+    let perms = DenoPermissions::from_flags(&DenoFlags {
+      read_whitelist: whitelist.clone(),
+      write_whitelist: whitelist.clone(),
+      no_prompts: true,
+      ..Default::default()
+    });
+
+    // Inside of /a/specific and /a/specific/dir/name
+    perms.check_read("/a/specific/dir/name").unwrap();
+    perms.check_write("/a/specific/dir/name").unwrap();
+
+    // Inside of /a/specific but outside of /a/specific/dir/name
+    perms.check_read("/a/specific/dir").unwrap();
+    perms.check_write("/a/specific/dir").unwrap();
+
+    // Inside of /a/specific and /a/specific/dir/name
+    perms.check_read("/a/specific/dir/name/inner").unwrap();
+    perms.check_write("/a/specific/dir/name/inner").unwrap();
+
+    // Inside of /a/specific but outside of /a/specific/dir/name
+    perms.check_read("/a/specific/other/dir").unwrap();
+    perms.check_write("/a/specific/other/dir").unwrap();
+
+    // Exact match with /b/c
+    perms.check_read("/b/c").unwrap();
+    perms.check_write("/b/c").unwrap();
+
+    // Sub path within /b/c
+    perms.check_read("/b/c/sub/path").unwrap();
+    perms.check_write("/b/c/sub/path").unwrap();
+
+    // Inside of /b but outside of /b/c
+    perms.check_read("/b/e").unwrap_err();
+    perms.check_write("/b/e").unwrap_err();
+
+    // Inside of /a but outside of /a/specific
+    perms.check_read("/a/b").unwrap_err();
+    perms.check_write("/a/b").unwrap_err();
+  }
+
+  #[test]
+  fn check_domains() {
+    let perms = DenoPermissions::from_flags(&DenoFlags {
+      net_whitelist: vec![
+        "localhost".to_string(),
+        "deno.land".to_string(),
+        "github.com:3000".to_string(),
+        "127.0.0.1".to_string(),
+        "172.16.0.2:8000".to_string(),
+      ].to_vec(),
+      no_prompts: true,
+      ..Default::default()
+    });
+
+    // http and https for the same domain should pass
+    perms.check_net("http://localhost/test/url").unwrap();
+    perms.check_net("https://localhost/test/url").unwrap();
```

Comment:
This seems very odd that it operates on URLs. I think it should only be passed a host.